### PR TITLE
fix cario x64 bit crash + apply version info to libpng16.dll

### PR DESCRIPTION
**Pull request by BingTSC**
(1) fix carios x64 bit crash
(2) apply version info to libpng16.dll

image

### CODE REVIEW

File: gvsbuild/projects.py [hunk 1]
```diff
@@ -70,6 +70,7 @@ def __init__(self):
             archive_url = 'http://cairographics.org/snapshots/cairo-1.15.2.tar.xz',
             hash = '268cc265a7f807403582f440643064bf52896556766890c8df7bad02d230f6c9',
             dependencies = ['fontconfig', 'glib', 'pixman', 'libpng'],
+			patches = ['cairo.patch']
```
Comment (TingPing) [inline gvsbuild/projects.py]: Incorrect whitespace.

Comment (BingTSC) [reply gvsbuild/projects.py]: fixed.

**Comment by nacho** [comment]
Can you create a different pull request for the png patch? it looks correct to me so I think it should be fine to put it in. Instead the one for cairo I think it is not completely correct, that will work in 64bit but I guess it will fail on 32bit

**Comment by BingTSC** [comment]
> I guess it will fail on 32bit

why? I think 32bit can also handle unsigned long long
https://stackoverflow.com/questions/330374/long-long-implementation-in-32-bit-machine

**Comment by BingTSC** [comment]
> how do you reproduce the crash?

our product has a crash before fixed by this patch file.  but I don't know how to reproduce outside.

> did you file a ticket for upstream cairo?

no

if you really concern, I can just take that commit away.

**Comment by nacho** [comment]
I am fine adding this patch here, I just want to understand how to reproduce it and how to get this upstream, otherwise next time we update cairo we might not be able to understand whether we have to keep the patch or if upstream fixed the issue.

**Comment by BingTSC** [comment]
sorry I cannot help more. I didn't use cairo outside our product.  The crash reason is due to pointer truncation. ( as mentioned in commit description, the pointer in x64 is 64bit, but unsigned long is 32bit. so the pointer has been truncated and causes crash. )  just let me know if you just want png fix.

**Comment by nacho** [comment]
I just spoke with the maintainer of cairo, any chance you could at least get us a backtrace of the crash you get without the patch you made?

**Comment by BingTSC** [comment]
@nacho I did found the old bug on my side, because it's closed long long time ago, I didn't see any crash dump still saved there.  I'm very sorry :-(  the bug says it only crashed on x64 not win32. after fixing, QA tested both works on x64 and win32.

**Comment by nacho** [comment]
@BingTSC I understand the patch works fine, and as I said I am fine adding it to gvsbuild, but in order to get it in cairo we need to understand how to reproduce it. Any chance you could remove the patch, reproduce the crash with your application and get us a backtrace?

**Comment by BingTSC** [comment]
@nacho  I was just trying but no luck.  I'm very sorry.  Next time I'll remember to save the crash dump for reference.

**Pull request merged**